File: src/libeq/nr.f
C FILE: nr.f
C Subroutines borrowed from Numerical Recipes
C
      SUBROUTINE ludcmp(a,n,indx,d)
      INTEGER n,np,indx(n),NMAX
      REAL d,a(n,n),TINY
Cf2py intent(inout) a
Cf2py intent(in) n, indx, NMAX, d
C     Largest expected n,  and  a small  number.
      PARAMETER (NMAX=500,TINY=1.0e-20)
C     Given a matrix a(1:n,1:n), with physical dimension np by np, this routine replaces it by
C     the LU decomposition of a rowwise permutation of itself. a and n
C     are input. a is  output, arranged  as  in  equation (2.3.14) above; indx(1:n)
C     is an output vector that records the row permutation effected by the partial pivoting;
C     d is output as +/-1 depending on whether the  number  of  row  interchanges  was  
C     even or odd, respectively. This routine is used in combination  with
C     lubksb to  solve linear  equations  or invert  a  matrix.
      INTEGER i,imax,j,k
      REAL aamax,dum,sum,vv(NMAX)
C     vv stores the  implicit  scaling  of each row.  
      d=1.
C     No  row interchanges  yet.
      do 12 i=1,n
C       Loop  over  rows  to  get  the  implicit  scaling  information.
        aamax=0.
        do 11 j=1,n
          if (abs(a(i,j)).gt.aamax) aamax=abs(a(i,j))
11      enddo
C       No nonzero largest element.
C       if (aamax.eq.0.) pause 'singular matrix in ludcmp'
        vv(i)=1./aamax
C       Save  the  scaling.
12    enddo
      do 19 j=1,n
C       This is the loop over columns of Crout's method.
        do 14 i=1,j-1
C         This  is equation (2.3.12) except for i = j.
          sum=a(i,j)
          do 13 k=1,i-1
            sum=sum-a(i,k)*a(k,j)
13        enddo
          a(i,j)=sum
14      enddo
      aamax=0.
C     Initialize  for the search for largest pivot element.
      do 16 i=j,n
C       This is i = j of equation (2.3.12) and i = j +1 :::N of  equation (2.3.13).
        sum=a(i,j)
        do 15 k=1,j-1
          sum=sum-a(i,k)*a(k,j)
15      enddo
        a(i,j)=sum
        dum=vv(i)*abs(sum)
C       Figure  of  merit  for  the pivot.
        if (dum.ge.aamax) then
C         Is it  better than  the  best so far?
          imax=i
          aamax=dum
        endif
16    enddo
      if (j.ne.imax)then
C       Do  we  need to  interchange rows?
        do 17 k=1,n
C         Yes,  do  so...
          dum=a(imax,k)
          a(imax,k)=a(j,k)
          a(j,k)=dum
17      enddo
        d=-d
C       ...and change the  parity  of d.
        vv(imax)=vv(j)
C       Also interchange the  scale factor.
      endif
      indx(j)=imax
      if(a(j,j).eq.0.)a(j,j)=TINY
C       If the pivot element is zero the matrix is singular (at least to the
C       precision of the algorithm). For some applications on singular matrices,
C       it is desirable to substitute TINY for  zero.
        if (j.ne.n) then
C         Now, finally, divide by the pivot element.
          dum=1./a(j,j)
          do 18 i=j+1,n
            a(i,j)=a(i,j)*dum
18        end do
        end if
19    end do
C     Go back for the next column in  the reduction.
      return
      END

C     Here is the routine for forward substitution and backsubstitution, implementing
C     equations  (2.3.6)  and  (2.3.7).
      SUBROUTINE lubksb(a,n,indx,b)
      INTEGER n,indx(n)
      REAL a(n,n),b(n)
Cf2py intent(in) a, n, np, indx
Cf2py intent(out) b
C     Solves the  set of n linear  equations AX = B. Here a is input,  not  as the  matrix
C     A but rather as its LU decomposition, determined by the routine ludcmp .
C     indx is input as the permutation vector returned by ludcmp.  b(1:n)
C     is input as the right-hand side vector B, and returns with the solution vector
C     X. a, n, np, and indx are not modi ed by this routine
C     and can be left in place for successive calls with  different right-hand sides
C     b.  This  routine takes into account the possibility that b
C     will begin with many zero elements, so it is efficient for use in matrix inversion.
      INTEGER i,ii,j,ll
      REAL sum
      ii=0
C     When ii is set to a positive value, it will become the in- dex of
C     the rest nonvanishing element of b. We now do the forward substitution,
C     equation (2.3.6). The only new wrinkle is to unscramble the permutation as we go.
      do 22 i=1,n
        ll=indx(i)
        sum=b(ll)
        b(ll)=b(i)
        if (ii.ne.0)then
          do 21 j=ii,i-1
          sum=sum-a(i,j)*b(j)
21      enddo
        else if (sum.ne.0.) then
          ii=i
C         A nonzero element was encountered, so from now on we will
C         have  to do  the sums  in  the  loop above.
        endif
        b(i)=sum
22    enddo
      do 24 i=n,1,-1
C       Now we do the backsubstitution, equation  (2.3.7).
        sum=b(i)
        do 23 j=i+1,n
          sum=sum-a(i,j)*b(j)
23      enddo
        b(i)=sum/a(i,i)
C       Store a  component of the solution vector X.
24    enddo
      
C     To  summarize,  this is the preferred  way  to solve the linear  set  of  equations
C     A x = b :
C     call ludcmp(a,n,np,indx,d)
C     call lubksb(a,n,np,indx,b)
C     The  answer x will be returned in b.  Your original matrix A will have been destroyed.
C     If you subsequently want to solve a set  of equations with the same
C     A but a different  right-hand side b ,  you  repeat only
C     call lubksb(a,n,np,indx,b)
C     not,  of course,  with the original matrix A ,  but with a and indx
C     as  were  already returned  from ludcmp
      
      END

      SUBROUTINE lnsrch(n,xold,fold,g,p,x,f,stpmax,check,func)
      INTEGER n
      LOGICAL check
      REAL f,fold,stpmax,g(n),p(n),x(n),xold(n),func,ALF,TOLX
Cf2py intent(in) n, xolf, fold, g, p, x, f, stpmax, check, func
Cf2py intent(out) g, x
      PARAMETER (ALF=1.e-4,TOLX=1.e-7)
      EXTERNAL func
C  USES func
C     Given an n-dimensional point xold(1:n), the value of the function and gradient there,
C     fold and g(1:n), and a direction p(1:n), nds a new point x(1:n) along the direction
C     p from xold where the function func has decreased sufficiently."The new function value
C     is returned in f. stpmax is an input quantity that limits the length of the steps so that you
C     do not try to evaluate the function in regions where it is  unde ned  or subject to overflow.
C     p is usually the Newton direction. The output quantity check
C     is false on a normal exit. It is true  when x is  too  close  to xold.
C     In a minimization  algorithm, this usually signals
C     convergence and can be  ignored.  However, in  a zero-finding algorithm the calling  program
C     should check  whether  the  convergence  is  spurious.
C     Parameters:
C      ALF       ensures  sufficient  decrease  in  function  value;
C      TOLX is  the  convergence criterion  on Delta x .
      INTEGER i
      REAL a,alam,alam2,alamin,b,disc,f2,rhs1,rhs2,slope,
     *     sum,temp,test,tmplam
      check=.false.
      sum=0.
      do 31 i=1,n
        sum=sum+p(i)*p(i)
31    enddo

      sum=sqrt(sum)
      if(sum.gt.stpmax)then
C       Scale if  attempted step is too big.
        do 32 i=1,n
          p(i)=p(i)*stpmax/sum
32      enddo
      endif
      slope=0.
      do 33 i=1,n
        slope=slope+g(i)*p(i)
33    enddo

C     if(slope.ge.0.) pause 'roundoff problem in lnsrch'
      test=0.
C     Compute lambda min.
      do 34 i=1,n
        temp=abs(p(i))/max(abs(xold(i)),1.)
        if(temp.gt.test)test=temp
34    enddo

      alamin=TOLX/test
      alam=1.
C     Always try full Newton step  rst.
1     continue
C       Start of iteration loop.
        do 35 i=1,n
          x(i)=xold(i)+alam*p(i)
35      enddo

        f=func(x)
        if(alam.lt.alamin)then
C         Convergence  on Delta x. For  zero finding, the calling program
C         should verify the convergence.
          do 36 i=1,n
            x(i)=xold(i)
36        enddo

          check=.true.
          return
      else if(f.le.fold+ALF*alam*slope)then
C        Sufficient function decrease.
         return
      else
C       Backtrack.
        if(alam.eq.1.)then
C         First  time.
          tmplam=-slope/(2.*(f-fold-slope))
        else
C         Subsequent backtracks.
          rhs1=f-fold-alam*slope
          rhs2=f2-fold-alam2*slope
          a=(rhs1/alam**2-rhs2/alam2**2)/(alam-alam2)
          b=(-alam2*rhs1/alam**2+alam*rhs2/alam2**2)/
     *      (alam-alam2)
          if(a.eq.0.)then
            tmplam=-slope/(2.*b)
          else
            disc=b*b-3.*a*slope
            if(disc.lt.0.)then
              tmplam=.5*alam
            else if(b.le.0.)then
              tmplam=(-b+sqrt(disc))/(3.*a)
            else
              tmplam=-slope/(b+sqrt(disc))
            endif
          endif
          if(tmplam.gt..5*alam)tmplam=.5*alam  ! lambda <= 0.5 lambda_1
        endif
      endif
      alam2=alam
      f2=f
      alam=max(tmplam,.1*alam) ! lambda >= 0.1 lambda_1
      goto 1  !   Try  again.
      END

File: src/libeq/nr.py
"""FILE: nr.py

Contains hard core numeric recipes.
"""

import numpy as np

import excepts
from libeq.cpluss import cpluss
from libeq.jacobian import jacobian
from libeq.fobj import fobj
from libeq.damping import damping as fdamping


__version__ = '0.2'


def NewtonRaphson(x0, beta, stoichiometry, T, max_iterations=1000,
                  threshold=1e-10, damping=True, forcer=True, scaling=False,
                  step_limiter=False, zero_offdiag=False, logc=False,
                  log=False, panic=True, mask_flag=False, **kwargs):
    r"""Solve the set of equations J·δc = -F using Newton-Raphson's method.

    Given an initial guess **x0** and the parameters of the titration, this
    function uses the Newton-Raphson method to solve the free
    concentrations.  It includes several
    convergence tecniques such as scaling, damping and step limiting. For the
    original source of damping and scaling see De Robertis, *et al.*,
    [#f1]_ and also Del Piero, *et al.* [#f2]_

    Parameters:
        x0 (:class:`numpy.ndarray`): initial guess for iterations. If *logc* is
            True, the natural logarithm of the concentrations is expected to be
            input. Otherwise, the regular concentration are expected.
        beta (:class:`numpy.ndarray`): values of equilibrium constants
        stoichiometry (:class:`numpy.ndarray`): stoichiometric coefficients
        T (:class:`numpy.ndarray`): total concentration values
        scaling (bool, optional): Whether to use or not scaling techniques.
            False by default. See :func:`DRscaling`.
        damping (bool, optional): Whether to use or not damping techniques.
            False by default. See :func:`DRdamping`.
        step_limiter (bool, optional[True]): Whether to use step limiter or
            not.  Step limiter attempts to avoid a step which leads to negative
            concentrations. It searches the increments which will result in
            negative concentrations and replaces the step
            :math:`x_n = x_{n-1} + dx` for
            :math:`x_n = x_{n+1} \cdot e^{dx}`
            which always gives a positive concentration as a result.
            It can jointly be used with the forcer.
        forcer (bool, optional[False]): Whether to use forcer techniques.
            False by default.
        threshold (float): threshold criterium for convergence
        max_iterations (int): maximum number of iterations allowed
        zero_offdiag (bool): If this option is applied, all non-diagonal
            elements of the jacobian are set to zero. This option is useful
            when only an estimate of the free concentrations is wanted.
        max_damps (int):  default, 2. maximum number of dampimgs allowed.
            It applies only if damping = True.
        logc (bool): Fit the logarithmic value of the concentration
        log (bool): The handle for logging
        do_iterations (int, optional): Perform exatly *do_iterations*
            iterations and do not test for convergence.
        panic (bool, optional, default True): If convergence fails, dump to a
            file the data to debug.
    Returns:
        :class:`numpy.ndarray`: An array containing the values of the free
            concentrations.
    Raises:
        :class:`RuntimeError`: if the concentrations cannot be calculated
        :class:`excepts.TooManyIterations`: if the maximum number of iterations
            is reached.

    .. warning:: This function is the very core of the program. Do not change
        the default parameters unless you know what you are doing.

    .. [#f1] *Analytica Chimica Acta* 1986, **191**, 385-398
    .. [#f2] Del Piero, *et al.*: *Annali di Chimica* 2006, 96.
    """

    def _panic_save():
        if panic:
            np.savez_compressed('consol_panic.npz', free_concentration=x0,
                                beta=beta, stoichiometry=stoichiometry,
                                analytc=T)

    if zero_offdiag and scaling:
        raise ValueError("Options scaling and zero_offdiag are not" +
                         "compatible with each other")

    if 'do_iterations' in kwargs:
        do_iterations = kwargs['do_iterations']
        if not isinstance(do_iterations, int):
            raise TypeError('do_iteration must be a positive int.')
        if do_iterations < 1:
            raise ValueError('do_iteration must be a positive int.')
    else:
        do_iterations = None

    iterations = 0
    n_species = stoichiometry.shape[1]

    # copy x0 so that it is not modified outside this function.
    x = np.copy(x0)
    # ------ check input --------
    # libaux.assert_array_dim(2, x0)
    x, T = np.atleast_2d(x, T)

    # ------ main loop ----------
    while True:
        if mask_flag and not logc:
            np.ma.masked_less_equal(x, 0.0, copy=False)

        c0 = cpluss(x, beta, stoichiometry, full=True, logc=logc)

        if logc:
            _c = np.exp(c0)
        else:
            _c = c0

        if mask_flag:
            F = np.ma.masked_invalid(fobj(_c, stoichiometry, T))
            J = np.ma.masked_invalid(jacobian(_c, stoichiometry, logc=logc))
            assert not np.any(np.isnan(J))
            assert not np.any(np.isnan(F))
        else:
            F = fobj(_c, stoichiometry, T)
            J = jacobian(_c, stoichiometry, logc=logc)
            msg = 'could not calculate %s (iteration %d)'
            if np.any(np.isnan(J)):
                _panic_save()
                msg2 = msg % ('jacobian', iterations)
                raise excepts.FailedCalculateConcentrations(msg2, x)
            if np.any(np.isnan(F)):
                msg2 = msg % ('residuals', iterations)
                raise excepts.FailedCalculateConcentrations(msg2, x)

        # TODO This should be deleted when debug is not necessary
        # if not iterations % 50:
        #     print('chisq(it:', iterations, 'n:', T.shape[0], ') = ',
        #           np.sum(F**2), np.max(np.abs(F)))

        if zero_offdiag:
            J *= np.eye(n_species)  # zerom

        if scaling:
            d = DRScaling(J, F)
            dx = np.linalg.solve(J, -F) / np.sqrt(d)
        else:
            dx = np.linalg.solve(J, -F)

        # np.nan_to_num(dx)

        if forcer:
            # TOLX = 1e-5
            # #!
            # tmp = x0/dx
            # tmp[np.isnan(tmp)] = 0.0
            # stpmax = np.max(-tmp, axis=1)
            # # stpmax = np.max(-x0/dx, axis=1)
            # dx *= stpmax[:, np.newaxis]/(1.0+TOLX)
            # slope = -np.sum(F**2, axis=1)
            # g2 = generate_data2(beta, stoichiometry, T, x, dx)
            l, _ = linesearch3(x, dx, beta, stoichiometry, T)
            # # l, _ = linesearch(x0, dx, slope, g2)
            x += l[:, None]*dx

            # ## linesearch1 ##
            # slope = -np.sum(F**2, axis=1)
            # g2 = generate_data2(beta, stoichiometry, T, x, dx)
            # l, _, x = linesearch1(x0, dx, slope, g2)
        else:
            if step_limiter:
                x = limit_step(x, dx)
            else:
                x += dx

        iterations += 1
        if do_iterations:
            if iterations >= do_iterations:
                break
        else:
            # -------- Test for convergence ---------
            # if np.all(np.abs(dx/x0) < threshold):
            if np.all(np.abs(F) < threshold):
                break

        if damping:
            fdamping(x, beta, stoichiometry, T)

        if iterations > max_iterations:
            raise excepts.TooManyIterations('too many iterations', x)

        if mask_flag:
            np.ma.masked_invalid(x, copy=False)
            np.ma.masked_less_equal(x, 0.0, copy=False)
            np.ma.mask_rows(x)

    return x


# def linesearch2(x0, dx, J0, F0, F1, step_limiter=True):
#     r"""2-point parabolic line search along Newton direction.
#
#     Implementation of 2-point parabolic linesearch. This is a variation
#     of the line search for 3 points.[#]_ This line search defines a function
#     :math:`f=\frac12F\cdot F` which is to be minimized. Then we define a
#     parameter λ that 0<λ<1 which is the fraction of the Newton step and
#     then another function *g* which is function of the fractional is defined
#     so that
#
#     .. math:: g(\lambda) = f(x_0 + \lambda \delta x)
#
#     We model :math:`g(\lambda)` as a parabolic function for which we know the
#     values of :math:`g(\lambda=0)` and :math:`g(\lambda=1)` as well as the
#     derivative :math:`g'(\lambda=0)=`
#
#     .. math:: \lambda = \frac{g_0'}{2(g_1-g_0-g_0')}
#
#     .. [#] W. H. Press, S. A. Teukolksy, W. T. Vetterling, Brian P. Flannery,
#        Numerical Recipes in C. The Art of Scientific Computing, Second
#        Edition 1997, pages 384--385.
#     """
#     slope = np.sum(np.sum(F0[..., None] * J0, axis=1)*dx, axis=1)
#     lam = np.full_like(slope, 0.1)
#     who = slope < 0.0
#     g0 = 0.5*np.sum(np.square(F0[who]), axis=1)
#     g1 = 0.5*np.sum(np.square(F1[who]), axis=1)
#     lam[who] = -0.5*slope[who] / (g1-g0-slope[who])
#     lam[lam > 1.0] = 1.0
#     if step_limiter:
#         return limit_step(x0, lam[:, None]*dx)
#     else:
#         return x0 + lam[:, None]*dx


def linesearch3(x0, dx, beta, stoichiometry, T, lmax=None, g0=None, g2=None):
    r"""Three-point parabolic line search.

    This functions implements a 3-point linesearch in the Newton direction.
    This is a variation of the line search for 2 points.[#]_ The function to
    be minimized is the same though but the approach is different and it is
    adapted to the nature of the problem of concentration solving. We define
    a function :math:`f=\frac12F\cdot F` which is to be minimized. Then we
    define a parameter λ that 0<λ<1 which is the fraction of the Newton step
    and then another function *g* which is function of the fractional is
    defined so that

    .. math:: g(\lambda) = f(x_0 + \lambda \delta x)

    We know that negative x₀ values are forbidden, therefore λ might limited
    to values lower than 1. The maximum value allowed for λ is that that makes
    any concentration equal to 0, therefore
    :math:`\lambda_{max} = -x_0/\delta` if :math:`-x_0/\delta<1`

    We model :math:`g(\lambda)` as a parabolic function for which we calculate
    the values for λ=0, λ=½λ(max) and λ=0.99λ(max).

    .. [#] W. H. Press, S. A. Teukolksy, W. T. Vetterling, Brian P. Flannery,
       Numerical Recipes in C. The Art of Scientific Computing, Second Edition
       1997, pages 384--385.
    """
    nerr = np.geterr()
    np.seterr(all='ignore')

    def g(l):
        "Auxiliary function."
        FF = fobj(cpluss(x0 + l[:, np.newaxis]*dx, beta, stoichiometry,
                         full=True),
                  stoichiometry, T)
        return 0.5*np.sum(np.square(FF), axis=1)

    if lmax is None:
        lmax = -x0/dx       # may cause division by 0
        # w = np.logical_or(lmax > 1.0, lmax < 0.0)
        # lmax[w] = 1.0
        lmax[lmax < 0.0] = 1.0
        lmax = np.min(lmax, axis=1)

    if g0 is None:
        g0 = g(np.zeros_like(lmax))

    g1 = g(lmax/2)
    x1 = x0+lmax[:, None]*dx
    x1[x1 < 0.0] = 0.0
    if g2 is None:
        g2 = g(0.99*lmax)

    b = (-g2+4*g1-3*g0)
    a = (g2-g0-b)/lmax
    lmin = -0.5*b/a         # may cause division by 0

    # In the unlikely case where a == 0.0 meaning g0 == g1 == g2 we set the
    # step halfway.
    w = a == 0.0
    if np.any(w):
        lmin[w] = lmax[w]/2

    w = lmin < 0.1*lmax     # Set minimum step as 0.1λ(max)
    lmin[w] = 0.1*lmax[w]

    w = lmin > lmax         # In the unlikely case where λ(min)>λ(max), set
    lmin[w] = 0.95*lmax[w]  # λ(min) close enough to λ(max)

    gmin = g(lmin)

    # Check g(lmin) < g0
    w2 = gmin > g0
    if beta.ndim == 1:
        _beta = beta
    else:
        _beta = beta[w2]
    if np.any(w2):
        lmin[w2], gmin[w2] = linesearch3(x0[w2], dx[w2], _beta, stoichiometry,
                                         T[w2], lmax=lmax[w2]/2, g0=g0[w2],
                                         g2=gmin[w2])
    np.seterr(**nerr)
    return lmin, gmin


def limit_step(x, dx):
    r"""Limit step.

    Given a state (**x**) and a step (**dx**), the next state is expected to be
    :math:`x+dx`. However, in some cases negative values are forbidden. This
    may happen for small values of the state. In the case of *x* being small
    we can approximate :math:`x+1 \simeq e^{-x}`

    Parameters:
        x (:class:`numpy.ndarray`): The state. It must be 1D.
        dx (:class:`numpy.ndarray`): The step. It must have  the same length
            than *x*.

    """
    if len(x) != len(dx):
        raise ValueError('both arguments must have the same size')
    who = (x + dx) > 0.0
    # return np.where(who, x+dx, x*np.exp(dx))
    return np.where(who, x+dx, x + (1 - np.exp(dx)))


def DRScaling(J, F):
    """Apply scaling to both jacobian and objective function.

    Applies scaling technique according to De Robertis, *et al.* [#f1]_
    The scaling technique overcomes the problem of divergence
    when the jacobian (here called G) matrix is near singular.
    "... scaling was applied to matrix *G* and to vector *e*
    (residuals, :math:`e = C_{k, calcd} - C_k`)
    according to the equations
    :math:`g_{kj}^* = g_{kj}(g_{kk}g_{jj})^{-1/2}`
    and
    :math:`e_k^* = e_kg_{kk}^{-1/2}` where :math:`g_{kj}^*`
    and :math:`e_k^*` are the elements of the scaled matrix and vector
    respectively."

    .. math:: J^*_{kj} = J_{kj}(J_{kk}J_{jj})^{-1/2}
    .. math:: F^*_{k} = F_{k}J_{kk}^{-1/2}

    Parameters:
        J (:class:`numpy.ndarray`): jacobian array, which will be modified.
            It can be of any dimensionality provided that the last two are
            of the same size.
        F (:class:`numpy.ndarray`): residuals array, which will be modified.
            If must have one dimmension less than J and the rest of the
            axes be of the same size than J.

    Returns:
        :class:`numpy.ndarray`: The diagonal of the jacobian, :math:`J_{ii}`
            to scale back the result.
    """
    # s = J.shape
    # i = np.indices(s)
    # assert s[-2] == s[-1]
    # d = (J[i[-2] == i[-1]]).reshape(s[:-1])
    d = J[np.arange(J.shape[0])[:, None], np.eye(J.shape[1], dtype=np.bool)]
    J /= np.sqrt(d[..., np.newaxis]*d[..., np.newaxis, :])
    F /= np.sqrt(d)
    return d


# DEPRECATED
def linesearch(x0, dx, slope, gfunc):
    if np.any(slope >= 0.0):
        raise ValueError('Negative slope values.')

    ALF = 1e-4
    TOLX = 1e-7

    fold = gfunc(np.zeros(dx.shape[0]))

    lmin = TOLX / np.max(np.absolute(dx/x0), axis=1)
    assert lmin.shape == (x0.shape[0], )

    lam = np.ones_like(lmin, dtype=np.float)
    lam2 = np.zeros_like(lam)
    f2 = np.zeros_like(fold)
    first = True

    while True:
        # x1 = x0+lam[:, None]*dx
        f = gfunc(lam)

        done_on_x = lam < lmin
        done_on_f = f < (fold+ALF*lam*slope)
        done = np.logical_or(done_on_x, done_on_f)
        if np.all(done):
            break

        if first:
            tmplam = -0.5*slope / (f-fold-slope)
            first = False
        else:
            rhs1 = f-fold-lam*slope
            rhs2 = f2-fold-lam2*slope
            a = (rhs1/lam**2-rhs2/lam2**2)/(lam-lam2)
            b = (-lam2*rhs1/lam**2+lam*rhs2/lam2**2)/(lam-lam2)

            w = conditions(~done, a == 0)
            tmplam[w] = -slope[w]/(2.*b[w])

            disc = b**2-3.*a*slope
            w = conditions(~done, disc < 0.0)
            tmplam[w] = 0.5*lam[w]

            w = conditions(~done, disc >= 0.0, b <= 0.0)
            tmplam[w] = (-b[w]+np.sqrt(disc[w]))/(3.*a[w])

            w = conditions(~done, disc >= 0.0, b > 0.0)
            tmplam[w] = -slope[w]/(b[w]+np.sqrt(disc[w]))

            w = conditions(~done, tmplam > 0.5*lam)
            tmplam[w] = 0.5*lam[w]

        lam2[~done] = lam[~done]
        f2[~done] = f[~done]
        lam[~done] = np.where(tmplam[~done] > 0.1*lam[~done],
                              tmplam[~done], 0.1*lam[~done])

    return lam, f


def generate_data2(beta, stoich, analyticalc, x0, dx):
    # From sandbox/linsrch.py
    def g(lam):
        x1 = x0+lam[:, None]*dx
        Fast = fobj(cpluss(x1, beta, stoich, full=True), stoich, analyticalc)
        gast = 0.5*np.sum(np.square(Fast), axis=1)
        return gast

    return g


def conditions(*args):
    # From sandbox/linsrch.py
    from functools import reduce
    return reduce(np.logical_and, args)


def linesearch1(x0, dx, slope, gfunc):
    if np.any(slope >= 0.0):
        raise ValueError('Negative slope values.')

    ALF = 1e-5
    TOLX = 1e-7

    g0 = gfunc(np.zeros(dx.shape[0]))

    lmin = TOLX / np.max(np.absolute(dx/x0), axis=1)
    assert lmin.shape == (x0.shape[0], )

    lam = np.ones_like(lmin, dtype=np.float)
    lam2 = np.zeros_like(lam)
    f2 = np.zeros_like(lam)
    first = True

    while True:
        g_lam = gfunc(lam)

        done_on_x = lam < lmin
        done_on_f = g_lam < (g0 + ALF*lam*slope)
        done = np.logical_or(done_on_x, done_on_f)
        if np.all(done):
            break

        _lam = lam[~done]
        _slope = slope[~done]
        _g_lam = g_lam[~done]
        _g0 = g0[~done]

        if first:
            tmplam = -0.5*_slope / (_g_lam - _g0 - _slope)
            first = False
        else:
            assert len(lam2[~done]) == len(f2[~done]) == len(_lam) == \
                   len(_slope) == len(_g0) == len(_g_lam)
            tmplam = minimum_poly3(_lam, lam2[~done], _g_lam, f2[~done], _g0,
                                   _slope)

        lam2 = lam[...]
        f2 = g_lam[...]
        import pudb
        pudb.set_trace()
        assert len(lam2[~done]) == len(f2[~done]) == len(_lam) == len(_slope)
        lam[~done] = np.where(tmplam > 0.1*_lam, tmplam, 0.1*_lam)

    assert np.all(np.logical_and(lam <= 1.0, lam >= 0.0))
    x_lam = x0 + lam[:, None]*dx
    return lam, g_lam, x_lam


def minimum_poly3(lam, lam2, f, f2, fold, slope):
    tmplam = np.empty_like(lam)
    rhs1 = f-fold-lam*slope
    rhs2 = f2-fold-lam2*slope
    a = (rhs1/lam**2-rhs2/lam2**2)/(lam-lam2)
    b = (-lam2*rhs1/lam**2+lam*rhs2/lam2**2)/(lam-lam2)

    w = (a == 0)
    tmplam[w] = -slope[w]/(2.*b[w])

    disc = b**2-3.*a*slope
    w = (disc < 0.0)
    tmplam[w] = 0.5*lam[w]

    w = np.logical_and(disc >= 0.0, b <= 0.0)
    tmplam[w] = (-b[w]+np.sqrt(disc[w]))/(3.*a[w])

    w = np.logical_and(disc >= 0.0, b > 0.0)
    tmplam[w] = -slope[w]/(b[w]+np.sqrt(disc[w]))

    w = (tmplam > 0.5*lam)
    tmplam[w] = 0.5*lam[w]
    return tmplam


def __mask(array):
    new_mask = collapse_mask_axis(array <= 0.0)
    np.ma.asarray(array)
    array.mask = new_mask


def collapse_mask_axis(array, axis=0):
    """Compact a multidimmensional mask into 1D mask.

    Given a mask array, it returns a 1D array based on one axis of it where
    :func:`numpy.any` is performed for every element of the axis in question.

    Parameters:
        array (:class:`numpy.ndarray`): Boolean mask array
        axis (int): the axis in which the operation is performed.

    Returns:
        :class:`numpy.ndarray`: the collapsed mask

    .. seealso:: :func:`mask_dim`
    """
    dims = tuple(i for i in range(array.ndim) if i != axis)
    return np.any(array, axis=dims)


def mask_dim(array, mask, axis):
    """Apply 1D-mask to a multidimmensional array along axis.

    Parameters:
        array (:class:`numpy.ndarray`): Boolean mask array
        mask (iterable): Boolean mask array
        axis (int): the axis in which the operation is performed.

    Returns:
        :class:`numpy.ndarray`: the collapsed mask
    """
    if array.shape[axis] != len(mask):
        raise ValueError

    shape1 = tuple(len(mask) if _ == axis else 1 for _ in range(array.ndim))
    shape2 = array.shape
    full_mask = np.broadcast_to(mask.reshape(shape1), shape2)
    return np.ma.array(array, mask=full_mask)
